### PR TITLE
feat: add CIDR range for NICs in `layer2`

### LINKED ISSUE #221: Unpredictable Node IP addresses
Hi,
my nodes have two interface, one public interface, for which I want to use my cloud providers CCM LB and one private interface for which I want to use OpenELB in layer 2 mode. According to the documentation I need to annotate my nodes with `layer2.porter.kubesphere.io/v1alpha1="<node_IP>"`.
However, due to cluster autoscaling, new nodes are constantly being added to and deleted from my cluster. The IP addresses of new nodes are not predictable in my environment. How would I go about?

### PR DESCRIPTION
**Pull request by amal-thundiyil**
## Description
Previously only IP of NIC could be added as annotation to node. With this PR, a CIDR format IP range can be used for identification of an NIC. All possible networks are iterated over in the CIDR range and selected accordingly.

## Related Links
- Fixes #221 
 
Signed-off-by: Amal Thundiyil <amal.s.thundiyil@gmail.com>

### CODE REVIEW

File: pkg/speaker/layer2/arp.go [hunk 1]
```diff
@@ -197,6 +204,34 @@ func (a *arpSpeaker) SetBalancer(ip string, nodes []corev1.Node) error {
 	return fmt.Errorf("node %s has no nexthop", nodes[0].Name)
 }
 
+func (a *arpSpeaker) setBalancerFromIPRange(cidr string) error {
+	var err error
+	// convert string to IPNet struct
+	_, ipv4Net, err := net.ParseCIDR(cidr)
+	if err != nil {
+		a.logger.Error(err, "layer2: failed to parse cidr from ip")
```
Comment (chaunceyjiang) [inline pkg/speaker/layer2/arp.go:212]: An error has occurred here. Why not return?

**Comment by chaunceyjiang** [comment]
Thank you for your work. I will test this PR later

File: pkg/speaker/layer2/arp.go [hunk 2]
```diff
@@ -197,6 +204,34 @@ func (a *arpSpeaker) SetBalancer(ip string, nodes []corev1.Node) error {
...
+	// loop through addresses as uint32
+	for i := start; i <= finish; i++ {
+		// convert back to net.IP
+		ip := make(net.IP, 4)
+		binary.BigEndian.PutUint32(ip, i)
+		err = a.setBalancer(ip.String(), []string{ip.String()})
+		if err == nil {
+			return err
+		}
+	}
```
Comment (renyunkang) [inline pkg/speaker/layer2/arp.go:231]: I failed the test. You should make sure the ip is the node of the cluster before calling `setbalancer`

Comment (renyunkang) [inline pkg/speaker/layer2/arp.go:231]: I failed the test. You should make sure the ip is the node of the cluster before calling `setbalancer`

File: pkg/speaker/layer2/arp.go [hunk 3]
```diff
@@ -197,6 +204,34 @@ func (a *arpSpeaker) SetBalancer(ip string, nodes []corev1.Node) error {
...
+	return err
```
Comment (renyunkang) [inline pkg/speaker/layer2/arp.go:232]: I failed the test. You should make sure the ip is the node of the cluster before calling setbalancer.

**Comment by chaunceyjiang** [comment]
kindly ping @amal-thundiyil

**Comment by amal-thundiyil** [comment]
@chaunceyjiang, is there anything you'd like me to change in this PR?

**Comment by chaunceyjiang** [comment]
@amal-thundiyil 

Could you please provide some test screenshots?

**Comment by amal-thundiyil** [comment]
> @amal-thundiyil
Could you please provide some test screenshots?

Sorry for the delayed response. I have added a test.

**Comment by chaunceyjiang** [comment]
Thanks!
/lgtm

**Comment by renyunkang** [comment]
Regarding the value of `openelb eip`, we still need to solve it in other ways. There is no problem with node labels specifying NICs using the CIDR range.
_(We use the network card specified by `eip` to listen for arp requests. Use `layer2.porter.kubesphere.io/v1alpha1` to specify the next hop, and the mac corresponding to the specified ip as the arp response packet.)_

Thanks !!!

/lgtm
/approve

**Pull request merged**